qiskit/transpiler/passes/routing/swap_strategies/swap_strategy_router.py

```diff
@@ -0,0 +1,336 @@
+
+    This pass manager relies on the ``current_layout`` which corresponds to the qubit layout as
+    swap gates are applied. The pass will traverse all nodes in the dag. If a node should be
+    routed using a swap strategy then it will be decomposed into sub-instructions with swap
+    layers in between and the ``current_layout`` will be modified. Nodes that should not be
+    routed using swap strategies will be added back to the dag taking the ``current_layout``
+    into account.
+    """
+
+    def __init__(self, swap_strategy: Optional[SwapStrategy] = None) -> None:
+        """
+        Args:
+            swap_strategy: An instance of a :class:`.SwapStrategy` that holds the swap layers
+                that are used, and the order in which to apply them, to map the instruction to
+                the hardware. If this field is not given if should be contained in the
+                property set of the pass. This allows other passes to determine the most
+                appropriate swap strategy at run-time.
+        """
+        super().__init__()
+        self._swap_strategy = swap_strategy
+        self._bit_indices = None
+
+    def run(self, dag: DAGCircuit) -> DAGCircuit:
+        """Run the pass by decomposing the nodes it applies on.
+
+        Args:
+            dag: The dag to which we will add swaps.
+
+        Returns:
+            A dag where swaps have been added for the intended gate type.
+
+        Raises:
+            TranspilerError: If the swap strategy was not given at init time and there is
+                no swap strategy in the property set.
+        """
+        if self._swap_strategy is None:
+            swap_strategy = self.property_set["swap_strategy"]
+
+            if swap_strategy is None:
+                raise TranspilerError("No swap strategy given at init or in the property set.")
+        else:
+            swap_strategy = self._swap_strategy
+
+        new_dag = dag.copy_empty_like()
+
+        current_layout = Layout.generate_trivial_layout(*dag.qregs.values())
+
+        # Used to keep track of nodes that do not decompose using swap strategies.
+        accumulator = new_dag.copy_empty_like()
+
+        self._bit_indices = {bit: index for index, bit in enumerate(dag.qubits)}
+
+        for node in dag.topological_op_nodes():
+            if isinstance(node.op, Commuting2qBlocks):
+
+                # Check that the swap strategy creates enough connectivity for the node.
+                self._check_edges(node, swap_strategy)
+
+                # Compose any accumulated non-swap strategy gates to the dag
+                accumulator = self._compose_non_swap_nodes(accumulator, current_layout, new_dag)
+
+                # Decompose the swap-strategy node and add to the dag.
+                new_dag.compose(self.swap_decompose(dag, node, current_layout, swap_strategy))
+            else:
+                accumulator.apply_operation_back(node.op, node.qargs, node.cargs)
```

Comment:
It seems the `SwapStrategyRouter` only does a correct routing when given a circuit with some commuting paulis in it? If another gate is in there it does not get routed correctly if I'm reading this code right.
In that case I think we need a more descriptive name since "swap strategy" doesn't really imply this. Something like a commuting pauli router.

Reply:
I added a comment to this effect. See here: [b7e3344](https://github.com/Qiskit/qiskit-terra/pull/7813/commits/b7e33447bd5b0fd2cd88335fe25b482573ed8586)